### PR TITLE
Eden SDN

### PR DESCRIPTION
**Pull request by milan-zededa**
This is a long-term project aimed to enhance Eden with a programmable networking in order to extend the scope of use-cases that EVE can be tested for wrt. device connectivity.
The proposal and design description can be found on lf-edge wiki: https://wiki.lfedge.org/pages/viewpage.action?pageId=60064091

TODO:
- [x] Build of SDN VM integrated into "eden setup"
- [x] Update of eden tests
- [x] Documentation
- [x] Examples
- [x] In-depth testing of eden CLIs

### CODE REVIEW

File: pkg/eden/qemu.go
Comment (giggsoff) [inline pkg/eden/qemu.go:108]: Current behavior is to use nets[0].SecondAddress for the second dhcpstart. Without this we may have the same addresses on EVE interfaces as I can see in my experiments. In that case I cannot connect to EVE and tests indicates that problem as well.

File: tests/workflow/testdata/deploy_docker_eden.txt
Comment (giggsoff) [inline tests/workflow/testdata/deploy_docker_eden.txt:33]: Do we plan to use new notation in all tests where we want to access an app? If so, I think we should decide what to do with GCP tests. We [set](https://github.com/milan-zededa/eden/blob/35a1f570b65d5f16c6b3a600e983968c2e506bc9/pkg/models/gcp.go#L30) `eve.remote` internally for GCP devmodel and new implementation will [return](https://github.com/milan-zededa/eden/blob/35a1f570b65d5f16c6b3a600e983968c2e506bc9/cmd/sdn.go#L422) error.

**Review by giggsoff — COMMENTED** [review]
Thank you for such great contribution!
It is impossible to go over all the edits at once, so the comments are somewhat scattered. For me, the biggest question is the performance of EdenGCP, which is quite difficult to test.

File: cmd/debug.go
Comment (milan-zededa) [reply cmd/debug.go]: I have improved this and now the sdnForward command is able to handle all 3 cases:
1. EVE running remotely
2. EVE running locally but without SDN
3. EVE running locally with SDN
This means that this code does not have to separate between these cases anymore. But the main motivation was eden tests, where we want to be able to use the same command for connectivity testing regardless of the config.

File: sdn/README.md
Comment (milan-zededa) [reply sdn/README.md:6]: This is actually markdown syntax for forced line-break (without adding a new paragraph).

File: sdn/build.yml
Comment (milan-zededa) [reply sdn/build.yml]: Yes it is planned, but first I wanted to have this approved before requesting a new repo.

Comment (milan-zededa) [reply sdn/build.yml]: Done, moved code under `vm` subfolder (best dir name I could come up with).

File: sdn/go.mod
Comment (milan-zededa) [reply sdn/go.mod]: Done

File: sdn/README.md
Comment (giggsoff) [reply sdn/README.md:6]: Ah, good to know, thank you.

File: sdn/vm.yml.in
Comment (milan-zededa) [reply sdn/vm.yml.in]: Fixed - I actually wanted SDN tag, not eden version.

File: .github/workflows/eden.yml
Comment (milan-zededa) [reply .github/workflows/eden.yml]: No, removed. Forgot to do so after disabling SDN for github workflows.

File: pkg/eden/qemu.go
Comment (milan-zededa) [reply pkg/eden/qemu.go]: Fixed

File: tests/workflow/testdata/deploy_docker_eden.txt
Comment (milan-zededa) [reply tests/workflow/testdata/deploy_docker_eden.txt:33]: Fixed, now the function can handle any supported setup.

File: pkg/eden/qemu.go
Comment (giggsoff) [reply pkg/eden/qemu.go]: Thank you!

File: cmd/sdn.go
Comment (giggsoff) [inline cmd/sdn.go:552]: log.Fatalf will call os.Exit internally and will not call deferred closeTunnel function, so it will be good to return error from the function and run Fatal from outside or to run closeTunnel here explicitly.

What is strange with tunneling (and why Eden tests in CI are failing on log_test) is the behavior on disconnection:
if I run `timeout 5 ./eden eve ssh sleep 10` without tunneling (sdn), I see (for example with `logread -f` from EVE-OS console:

```
2022-10-08T13:04:51Z,debug;Accepted publickey for root from 192.168.0.2 port 56602 ssh2: RSA SHA256:Si10Cpt9etzNitsIuDBl68C1zvJ+aGZcH3D05AlHc1A
2022-10-08T13:04:56Z,debug;Received disconnect from 192.168.0.2 port 56602:11: disconnected by user
2022-10-08T13:04:56Z,debug;Disconnected from user root 192.168.0.2 port 56602
```

but with sdn I can see only:
```
2022-10-08T13:06:18Z,debug;Accepted publickey for root from 240.0.0.2 port 39160 ssh2: RSA SHA256:TRLMCoylLTiBG+dGgwz0YIdC97GNupjbLW89bTsxRWw
```

WDYT, is it expected (in that case we should modify [log_test](https://github.com/lf-edge/eden/blob/85527ca4614620b934b2dc63cdbbda72a6a631d2/tests/lim/testdata/log_test.txt#L7)) or we do not close some sessions?

File: pkg/utils/config.go
Comment (giggsoff) [inline pkg/utils/config.go:472]: Let's make it optional (true by default) until we stabilize Eden tests workflow and merge.

File: cmd/sdn.go
Comment (milan-zededa) [reply cmd/sdn.go]: > log.Fatalf will call os.Exit internally and will not call deferred closeTunnel function, so it will be good to return error from the function and run Fatal from outside or to run closeTunnel here explicitly.

Fixed - the function now returns error and Fatal is called from outside.

> WDYT, is it expected (in that case we should modify [log_test](https://github.com/lf-edge/eden/blob/85527ca4614620b934b2dc63cdbbda72a6a631d2/tests/lim/testdata/log_test.txt#L7)) or we do not close some sessions?

This is quite a mystery for me right now. In both cases (SDN and non-SDN) I can see full TCP termination sequence (FINs+ACKs), yet in SDN case the SSH daemon doesn't report the client being disconnected. I will continue investigating...

File: pkg/utils/config.go
Comment (milan-zededa) [reply pkg/utils/config.go]: Done - disabled for now.

**Review by giggsoff — APPROVED** [review]

File: sdn/build.yml
Comment (milan-zededa) [reply sdn/build.yml]: Sent request to Linux Foundation to get the dockerhub repo created.

Comment (giggsoff) [reply sdn/build.yml]: Great, thank you!

**Pull request merged**
